File: IdleClicker/Services/StorageService.swift
import Foundation

class StorageService {
    private let gameStateKey = "com.idleclicker.gamestate"

    func saveGameState(_ state: GameState) {
        do {
            let data = try JSONEncoder().encode(state)
            UserDefaults.standard.set(data, forKey: gameStateKey)
        } catch {
            print("Failed to save game state: \(error)")
        }
    }

    func loadGameState() -> GameState? {
        guard let data = UserDefaults.standard.data(forKey: gameStateKey) else {
            return nil
        }

        do {
            let state = try JSONDecoder().decode(GameState.self, from: data)
            return state
        } catch {
            print("Failed to load game state: \(error)")
            return nil
        }
    }

    func resetGameState() {
        UserDefaults.standard.removeObject(forKey: gameStateKey)
    }
}

File: idle_clicker_flutter/lib/services/storage_service.dart
import 'package:shared_preferences/shared_preferences.dart';
import '../models/game_state.dart';

class StorageService {
  static const String _gameStateKey = 'com.idleclicker.gamestate';

  Future<void> saveGameState(GameState state) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_gameStateKey, state.toJsonString());
  }

  Future<GameState?> loadGameState() async {
    final prefs = await SharedPreferences.getInstance();
    final jsonString = prefs.getString(_gameStateKey);
    if (jsonString == null) return null;
    try {
      return GameState.fromJsonString(jsonString);
    } catch (e) {
      print('Error loading game state: $e');
      return null;
    }
  }

  Future<void> resetGameState() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_gameStateKey);
  }
}
